Add the ability to migrate CoreDNS configmap in kubeadm

rajansandeep opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
Currently in kubeadm, when upgrading from Kubernetes (Eg. k8s 1.13 -> 1.14), the CoreDNS ConfigMap is retained as-is and applied to the upgraded version. This is done due to the possibility that the user could have customized the Corefile for their use-case. 

This approach is disadvantageous because this can lead to an incompatible Corefile which can lead to CoreDNS entering a `CrashLoopBackOff` state, and hence DNS not working in the cluster.

This PR adds the ability for users to seamlessly migrate their CoreDNS Configuration when they upgrade their clusters using kubeadm. The [CoreDNS Migration library](https://github.com/coredns/deployment/tree/master/kubernetes/migration) helps to handle migrations of CoreDNS Corefiles to be compatible with new versions of CoreDNS. 

The PR includes 3 preflight-checks while upgrading to check:
- If the CoreDNS image is an official release
- If the Migration of CoreDNS is possible (Checks if there are any errors while migrating the CoreDNS Config from the current version to the next version)
- If there are any CoreDNS plugins present that are unsupported by the migration library.

The user will be provided Logs and can override these checks.

During Upgrade:
- There is a check to see if the CoreDNS Config is the default (not edited). In that case, the user will be upgraded to the new default CoreDNS Config.
- If the CoreDNS Config has been modified, then the user will be migrated to the equivalent compatible Config for the latest version of CoreDNS. This will include handling all deprecations that may be present.

The user will be informed appropriately through logs.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:
The first commit includes the code change required for this PR.
The second commit includes the vendor and go mod changes.

This PR also updates the CoreDNS version to 1.5.0
**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Kubeadm now seamlessly migrates the CoreDNS Configuration when upgrading CoreDNS.
```

rajansandeep committed:
add the ability to migrate coredns configmap

rajansandeep commented:
/cc @chrisohaver @neolit123

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/common.go:
```diff
@@ -92,6 +93,10 @@ func enforceRequirements(flags *applyPlanFlags, dryRun bool, newK8sVersion strin
 		return nil, nil, nil, errors.Wrapf(err, "couldn't create a Kubernetes client from file %q", flags.kubeConfigPath)
 	}
 
+	if err := runCoreDNSpreflightChecks(client, ignorePreflightErrorsSet); err != nil {
```
the upgrade checks should have a single entry point: `runPreflightChecks`
the call to `runCoreDNSpreflightChecks()` should be moved there.
`client` needs to be passed to `runPreflightChecks`.

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/common.go:
```diff
@@ -181,6 +186,19 @@ func runPreflightChecks(ignorePreflightErrors sets.String) error {
+		return err
+	}
+	if currentDNSType == kubeadmapi.CoreDNS {
+		fmt.Println("[preflight] Running CoreDNS migration pre-flight checks.")
```
we can omit this notification, the preflight package will show check names.

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/common.go:
```diff
@@ -181,6 +186,19 @@ func runPreflightChecks(ignorePreflightErrors sets.String) error {
+	}
+	if currentDNSType == kubeadmapi.CoreDNS {
+		fmt.Println("[preflight] Running CoreDNS migration pre-flight checks.")
+		return upgrade.RunCoreDNSMigrationCheck(client, ignorePreflightErrors)
```
```
currentDNSType, _, err := dns.DeployedDNSAddon(client)
	if err != nil {
		return err
	}
	if currentDNSType == kubeadmapi.CoreDNS {
...
```

can be inside `upgrade.RunCoreDNSMigrationCheck`, and optionally make the function a no-op if kube-dns is used.
(i.e. exit early and don't run `preflight.RunChecks(migrationChecks, os.Stderr, ignorePreflightErrors)`)

neolit123 commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -297,6 +298,83 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
+		}
+	}
+	// Get the Corefile and installed CoreDNS version.
+	corefile, currentInstalledCoreDNSVersion, err := GetCoreDNSInfo(client)
```
`GetCoreDNSInfo()` can accept a `*ConfigMap` as an argument.
if `nil` it should fetch the CM, if non-`nil` it should use the given CM.
this avoids fetching the same CM twice.

neolit123 commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -297,6 +298,83 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
+	if !migration.Default("", corefile) {
+		updatedCorefile, err := migration.Migrate(currentInstalledCoreDNSVersion, kubeadmconstants.CoreDNSVersion, corefile, false)
+		if err != nil {
+			return errors.Wrap(err, "unable to migrate CoreDNS Configmap")
```
`Configmap` -> `ConfigMap`

neolit123 commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -297,6 +298,83 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
+				"Corefile-previous": corefile,
+			},
+		}); err != nil {
+			return errors.Wrap(err, "unable to update configmap")
```
`the CoreDNS ConfigMap`

neolit123 commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -297,6 +298,83 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
+			},
+			Data: map[string]string{
+				"Corefile":          updatedCorefile,
+				"Corefile-previous": corefile,
```
the Corefile itself is not that much data, but i wonder if this is the best backup option. the alternative is to write to file, which seems slightly better to me, but the question is where.

i'd like to hear more opinions on this.

chrisohaver replied:
> the Corefile itself is not that much data, but i wonder if this is the best backup option. the alternative is to write to file, which seems slightly better to me, but the question is where.
> i'd like to hear more opinions on this.

FWIW, my biased opinion ... Keeping the old version "in kubernetes" (e.g. in the configmap) seems more intuitive to me, from a user perspective. The old Corefile is readily available to roll back to, or to see as reference in the event that the migrated file is not working as expected.  In the ConfigMap, it's already in a place that any user that has permission to edit the configmap can access (same configmap).  If saved to a file (i assume to the local filesystem), it would only be saved to where ever kubeadm was executed.

johnbelamaric replied:
Yes, I agree completely with @chrisohaver

liggitt replied:
> The old Corefile is readily available to roll back to

does kubeadm do anything to rollback coredns versions? if so, does it also handle putting the "previous" data back as the current config? the "previous" key seems underspecified to be usable in an automated process... nothing indicates what version it is intended to work with. If I'm reading this correctly, even patch release upgrades would move current config into the previous key, right?

neolit123 commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -297,6 +298,83 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
 
+func createCoreDNSConfigMap(client clientset.Interface, cm *v1.ConfigMap) error {
+	// Get the CoreDNS ConfigMap. If not present, Create a new default ConfigMap.
+	if _, err := client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Get(kubeadmconstants.CoreDNSConfigMap, metav1.GetOptions{}); err != nil {
```
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/apiclient/idempotency.go#L96
can `CreateOrRetainConfigMap` be reused here?

neolit123 commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -297,6 +298,83 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
+		}
+		changes, err := migration.Deprecated(currentInstalledCoreDNSVersion, kubeadmconstants.CoreDNSVersion, corefile)
+		if err != nil {
+			return errors.Wrap(err, "unable to get list of notifications")
```
possibly use the naming "changes" in this error too?
optionally clarify more what action actually failed.

neolit123 commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -297,6 +298,83 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
+		if err != nil {
+			return errors.Wrap(err, "unable to get list of notifications")
+		}
+		klog.Infof("the CoreDNS Configuration has been migrated and applied: %v.", updatedCorefile)
```
`Configuration->configuration`

i think we should do the following here:
```
fmt.Pritnln("Migrating CoreDNS Corefile..."
changes, err := migration.Deprecated(curr......
...
// show the migration changes
klog.V(2).Infof(....)
klog.V(2).Infof(....)
klog.V(2).Infof(....)
...
```

neolit123 commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -297,6 +298,83 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
+		}
+	}
+	// If the current Configuration present is the default (unmodified) version, just update it to the latest default.
+	if migration.Default("", corefile) {
```
this check should be on top of:
`if !migration.Default("", corefile) {`

if it passes the function should `return nil` - i.e. exit early.

neolit123 commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -297,6 +298,83 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
+	// If the current Configuration present is the default (unmodified) version, just update it to the latest default.
+	if migration.Default("", corefile) {
+		if _, err := client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Update(cm); err != nil {
+			return errors.Wrap(err, "unable to update configmap")
```
`the CoreDNS ConfigMap`

neolit123 commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -297,6 +298,83 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
+	if err != nil && !apierrors.IsNotFound(err) {
+		return "", "", err
+	}
+	corefile := coreDNSConfigMap.Data["Corefile"]
```
i'd still check if the key exists here.
if the user has played with the CM and deleted it, this will panic.

neolit123 commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -297,6 +298,83 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
+	}
+	corefile := coreDNSConfigMap.Data["Corefile"]
+
+	dnsType, currentCoreDNSversion, err := DeployedDNSAddon(client)
```
i think it is possible to avoid checking if the addon type is CoreDNS in `GetCoreDNSInfo`.
the function should be called only if the addon is known to be CoreDNS.

neolit123 commented on cmd/kubeadm/app/phases/upgrade/migrate.go:
```diff
@@ -0,0 +1,135 @@
+/*
```
i think we should name this file `preflight.go` if are going to keep it under upgrade.
the alternative location is `cmd/kubeadm/app/preflight`, but i'm leaning towards leaving it in this package.

possibly needs discussion.

neolit123 commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -297,6 +298,83 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
 	return nil
 }
 
+func createCoreDNSConfigMap(client clientset.Interface, cm *v1.ConfigMap) error {
```
can we get an unit test for this function?

neolit123 commented on cmd/kubeadm/app/phases/upgrade/migrate.go:
```diff
@@ -0,0 +1,135 @@
+// checkUnsupportedPlugins checks if there are any plugins included in the current configuration
+// that is unsupported for migration.
+func checkUnsupportedPlugins(client clientset.Interface) error {
+	klog.V(1).Infoln("validating if there are any unsupported CoreDNS plugins in the Config")
```
`Config->Corefile`?

neolit123 commented on cmd/kubeadm/app/phases/upgrade/migrate.go:
```diff
@@ -0,0 +1,135 @@
+
+		}
+		if UnsupportedPlugins != "" || UnsupportedVersion != "" {
+			return errors.Errorf("there are unsupported plugins in the CoreDNS Configuration")
```
`Configuration->Corefile`?

neolit123 commented on cmd/kubeadm/app/phases/upgrade/migrate.go:
```diff
@@ -0,0 +1,135 @@
+
+// Checks if the CoreDNS image currently installed is an official released version.
+func checkRelease(client clientset.Interface) error {
+	klog.V(1).Infoln("validating if the image installed is the  official release of CoreDNS.")
```
extra space before `official`.

neolit123 commented on cmd/kubeadm/app/phases/upgrade/migrate.go:
```diff
@@ -0,0 +1,135 @@
+
+	isValid := migration.Released(coreDNSImageSha[1])
+	if !isValid {
+		return errors.Errorf("the CoreDNS image currently installed is not an official image. If you still wish to proceed, do so at your own risk.")
```
this is not a check we have for the control plane or etcd images.
to my understanding, if someone has built coredns from source at any SHA, `migration.Released` would return `false`. while this is still not recommended, blocking on it with a preflight check might be a bit too strict.

but possibly needs discussion / feedback from more people.

rosti replied:
+1 for not having this. Users in China may be implicated.

chrisohaver replied:
I think it is generally safe to remove this check.
  
The main reason for adding this check is to prevent external plugin users from inadvertently  replacing their custom CoreDNS with a stock one. A stock CoreDNS will crash when it encounters an unknown plugin in the Corefile at start up.

But most users just use a stock version of CoreDNS.  And IIUC, those who use CoreDNS compiled with external plugins will upgrade using "phases", and manually upgrade CoreDNS with an updated version compiled with the required external plugins if desired.

neolit123 reviewed (commented):
thanks for the PR @rajansandeep 
i understand the motivation of this change thus i've spent time doing a detailed review.

we might need at least some unit tests and possibly extend the release note that preflight checks are run now.

ideally more eyes from @kubernetes/sig-cluster-lifecycle-pr-reviews are needed here.

cblecker commented:
Is this needed/intended to merge before 1.15 code freeze?

rajansandeep commented:
@cblecker Yes! This is needed for 1.15

cblecker commented:
@rajansandeep Looks like the caddy dependency needs to be updated.

Can you please run:

```
hack/pin-dependency.sh github.com/mholt/caddy v1.0.0
hack/pin-dependency.sh github.com/BurntSushi/toml v0.3.1
hack/pin-dependency.sh github.com/golang/mock v1.2.0
hack/pin-dependency.sh github.com/google/uuid v1.1.1
hack/pin-dependency.sh github.com/gorilla/websocket v1.4.0
hack/pin-dependency.sh github.com/miekg/dns v1.1.3
hack/pin-dependency.sh github.com/onsi/ginkgo v1.8.0
hack/pin-dependency.sh github.com/onsi/gomega v1.5.0
hack/pin-dependency.sh github.com/russross/blackfriday v0.0.0-20170610170232-067529f716f4
hack/pin-dependency.sh github.com/stretchr/testify v1.3.0
hack/pin-dependency.sh gopkg.in/natefinch/lumberjack.v2 v2.0.0
hack/pin-dependency.sh gopkg.in/square/go-jose.v2 v2.2.2
hack/pin-dependency.sh gopkg.in/yaml.v2 v2.2.2
GO111MODULE=on go mod edit -dropreplace github.com/natefinch/lumberjack
GO111MODULE=on go mod edit -dropreplace github.com/shurcooL/sanitized_anchor_name
GO111MODULE=on go mod edit -dropreplace gopkg.in/yaml.v1
hack/update-vendor.sh
```

And commit the changes? Then when complete, include a comment with the output of `hack/lint-dependencies.sh`

neolit123 commented:
> Can you please run:

looks simple enough :roll_eyes: 

/priority important-soon
/milestone v1.15

rosti commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -297,6 +298,87 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
+		}
+		// show the migration changes
+		klog.V(2).Infof("the CoreDNS configuration has been migrated and applied: %v.", updatedCorefile)
+		klog.V(2).Infof("the old migration has been saved in the CoreDNS ConfigMap under the name [Corefile-previous]")
```
s/Infof/Infoln

rosti commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -297,6 +298,87 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
+		// show the migration changes
+		klog.V(2).Infof("the CoreDNS configuration has been migrated and applied: %v.", updatedCorefile)
+		klog.V(2).Infof("the old migration has been saved in the CoreDNS ConfigMap under the name [Corefile-previous]")
+		klog.V(2).Infof("The changes in the new CoreDNS Configuration are as follows:")
```
s/Infof/Infoln

rosti commented on cmd/kubeadm/app/phases/upgrade/migrate.go:
```diff
@@ -0,0 +1,143 @@
+
+		}
+		if UnsupportedPlugins != "" || UnsupportedVersion != "" {
+			return errors.Errorf("there are unsupported plugins in the CoreDNS Corefile")
```
errors.New

rosti commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -297,6 +298,87 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
+
+	// Check if the current Configuration present is the default (unmodified) version.
+	// If not, try and migrate the Config.
+	if !migration.Default("", corefile) {
```
Make this an `else` to the previous `if` statement?

rosti commented on cmd/kubeadm/app/phases/upgrade/migrate.go:
```diff
@@ -0,0 +1,143 @@
+
+// RunCoreDNSMigrationCheck initializes checks related to CoreDNS migration.
+func RunCoreDNSMigrationCheck(client clientset.Interface, ignorePreflightErrors sets.String) error {
+	currentDNSType, _, err := dns.DeployedDNSAddon(client)
```
This one is also present in the `ClusterConfiguration` which is already fetched from the config map at this point.

rosti commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -297,6 +298,87 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
+	}
+	corefile, ok := coreDNSConfigMap.Data["Corefile"]
+	if !ok {
+		return "", "", errors.Errorf("Unable to find the CoreDNS Corefile data.")
```
`errors.New`

rosti reviewed (commented):
Thanks @rajansandeep !
My main concern are the missing tests.

I also wonder if you are tracking the cluster addons effort?

chrisohaver commented:
> I also wonder if you are tracking the cluster addons effort?

@rosti, I have not heard of the cluster addons effort. Is there an open Issue/PR/KEP you can point to?

rosti commented:
Cluster Addons [meeting notes](https://docs.google.com/document/d/10_tl_SXcFGb-2109QpcFVrdrfnVEuQ05MBrXtasB0vk/edit), [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/addons/0035-20190128-addons-via-operators.md) and [repo](https://github.com/kubernetes-sigs/addon-operators).

In short, this effort is about producing cluster addon operators that can deploy, upgrade, manage, etc. different cluster addons (such as CoreDNS). It's pretty early on and so far it's mainly talks, however there are a couple of opened PRs in that repo that try to prototype an addon operator for CoreDNS.

cblecker commented:
@rajansandeep Please note that code freeze is EOD 5/30. Please address the comments and get things updated, or this may need to wait until 1.16.

neolit123 commented:
we had a discussion about this change during the kubeadm office hours today.
/assign @fabriziopandini 
for the final weight-in.

rajansandeep commented:
> Can you please run:

@cblecker I ran the vendor updates as instructed.

This is the output as requested of `hack/lint-dependencies.sh`:

```
$ hack/lint-dependencies.sh
These modules are pinned to versions different than the minimal preferred version.
That means that without require directives, a different version would be selected.
The command to switch to the minimal preferred version is listed for each module.

golang.org/x/crypto
    pinned:    v0.0.0-20181025213731-e84da0312774
    preferred: v0.0.0-20190228161510-8dd112bcdc25
    hack/pin-dependency.sh golang.org/x/crypto v0.0.0-20190228161510-8dd112bcdc25
golang.org/x/net
    pinned:    v0.0.0-20190206173232-65e2d4e15006
    preferred: v0.0.0-20190328230028-74de082e2cca
    hack/pin-dependency.sh golang.org/x/net v0.0.0-20190328230028-74de082e2cca
golang.org/x/sync
    pinned:    v0.0.0-20181108010431-42b317875d0f
    preferred: v0.0.0-20181221193216-37e7f081c4d4
    hack/pin-dependency.sh golang.org/x/sync v0.0.0-20181221193216-37e7f081c4d4
golang.org/x/sys
    pinned:    v0.0.0-20190209173611-3b5209105503
    preferred: v0.0.0-20190312061237-fead79001313
    hack/pin-dependency.sh golang.org/x/sys v0.0.0-20190312061237-fead79001313
golang.org/x/tools
    pinned:    v0.0.0-20190313210603-aa82965741a9
    preferred: v0.0.0-20190328211700-ab21143f2384
    hack/pin-dependency.sh golang.org/x/tools v0.0.0-20190328211700-ab21143f2384

Pinned module versions that aren't actually used:
GO111MODULE=on go mod edit -dropreplace github.com/natefinch/lumberjack
GO111MODULE=on go mod edit -dropreplace github.com/shurcooL/sanitized_anchor_name
GO111MODULE=on go mod edit -dropreplace gopkg.in/yaml.v1
```

rajansandeep commented:
@neolit123 @rosti I have cleaned up the code and added a unit test as requested
PTAL and let me know if more changes are needed or if I've missed something.

cblecker commented:
@rajansandeep Can you run the last commands too? Looks like something didn't take:
```
GO111MODULE=on go mod edit -dropreplace github.com/natefinch/lumberjack
GO111MODULE=on go mod edit -dropreplace github.com/shurcooL/sanitized_anchor_name
GO111MODULE=on go mod edit -dropreplace gopkg.in/yaml.v1
hack/update-vendor.sh
```

dstrebel commented:
Hello! We are starting Code Freeze for 1.15. Is this PR still planned to be included in 1.15?

fabriziopandini commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -350,7 +433,7 @@ func translateStubDomainOfKubeDNSToForwardCoreDNS(dataField string, kubeDNSConfi
 // in the form of Proxy for the CoreDNS Corefile.
 func translateUpstreamNameServerOfKubeDNSToUpstreamForwardCoreDNS(dataField string, kubeDNSConfigMap *v1.ConfigMap) (string, error) {
 	if kubeDNSConfigMap == nil {
```
This change seems unrelated to the corefile migration.
What is the rational behind this change?

rajansandeep replied:
When there is no `kube-dns ConfigMap`, the function should return `/etc/resolv.conf`. Else, the CoreDNS Corefile will break with no resolver to the `forward` plugin.
I found this bug while manually testing the upgrades. Probably didn't cause an issue previously because the CoreDNS ConfigMap was always retained and not updated when it already existed.

rosti replied:
I wonder if this can cause issues with `systemd-resolved` enabled systems (like recent Ubuntu versions)? Have you tried  this @rajansandeep ?

chrisohaver replied:
@rosti, IIUC, kubeadm already detects `systemd-resolved` systems, and configures `kubelet` to use an alternate `resolv.conf`.  From the perspective of Pods however, the path to this alternate `resolv.conf` is always `/etc/resolv.conf`, regardless of the original path.

fabriziopandini reviewed (approved):
@rajansandeep thanks for this PR!
@neolit123 @rosti thanks for helping in shaping this out nicely
I'm ok in trying to get this in last second, because having in place corefile migration allows kubeadm/users a better experience in upgrading CoreDNS 
/approve
/lgtm

Feel free to ping me in slack if I can help somehow to get it in this now

neolit123 commented:
@dstrebel yes.
@rajansandeep please rebase.

johnbelamaric commented:
@rajansandeep Why so many vendor changes? I think those need to be cleaned up.

rajansandeep commented:
@johnbelamaric, the original vendor changes are in [this commit](https://github.com/kubernetes/kubernetes/pull/78033/commits/e151d73d1f617beb765f1a3cd4984ffc00545555). @cblecker [suggested to update caddy deps](https://github.com/kubernetes/kubernetes/pull/78033#issuecomment-496747406), hence the extra vendor changes.

johnbelamaric commented:
But why do we need all that just for munging the Corefile? Do we really need all of caddy? Is there no go mod equivalent of dep prune?

neolit123 commented:
the caddyfile dependency is 100 lines of code:
https://github.com/mholt/caddy/blob/master/caddyfile/json.go#L106-L198

last time i checked kubeadm seems to be the only user, so i suggest that this is just copy paste this logic in the kubeadm dns.go file (uses APACHE2.0)

we had problems in the past, where caddy breaks and we have to ping maintainers.

neolit123 commented:
i don't understand why the vendor update blew up so much.
does caddy depend on ginkgo or miekg/dns?

chrisohaver commented:
> we had problems in the past, where caddy breaks and we have to ping maintainers.

Doesn't vendoring/go mod protect against that?

neolit123 commented:
> Doesn't vendoring/go mod protect against that?

it was before go modules.

chrisohaver commented:
> > Doesn't vendoring/go mod protect against that?
> 
> it was before go modules.

but not before vendoring?

neolit123 commented:
> but not before vendoring?

no, it broke verify-godep: https://github.com/mholt/caddy/pull/2345#issuecomment-439652315

cblecker commented:
/test all

chrisohaver commented:
I see, well I think this PR depends on different parts of caddy than the prior dependencies.
Nevertheless, we'd have to refactor the migration library to use local copies of whatever caddy functions are used, which I don't think we will have time to do before code freeze.

We _do_ have time to roll CoreDNS back to 1.3.1, if thats what we need to do at this point.  Not the end of the world.

Thoughts?

neolit123 commented:
> We do have time to roll CoreDNS back to 1.3.1, if thats what we need to do at this point. Not the end of the world.
> Thoughts?

EOD PST (code freeze) is in few hours.

cblecker commented:
From a dependency review standpoint, the vendor changes aren't actually highly objectionable. There's one brand new direct dependency being introduced with the migration library, and the other big piece is upgrading caddy to the stable tag (which brings in their telemetry package).

johnbelamaric commented:
If we only import github.com/mholt/caddy/caddyfile, can we just pull that in, instead of all of Caddy? It only imports standard libraries.

chrisohaver commented:
> If we only import github.com/mholt/caddy/caddyfile, can we just pull that in, instead of all of Caddy? It only imports standard libraries.

I don't really know how go modules work, but I think a go.mod has to be defined in the caddyfile directory for that to be possible (and there isn't).

johnbelamaric commented:
Ok, well it looks like it is using caddy.NewTestController anyway right now.

cblecker commented:
@johnbelamaric No, parse.go in caddyfile imports the telemetry package. That's why that is being pulled in. We aren't actually pulling in all of caddy.

chrisohaver commented:
> Ok, well it looks like it is using caddy.NewTestController anyway right now.

I could refactor the underlying library to use different parts, or no parts of caddy.  But cant do that before the code freeze at this point.

neolit123 commented:
looking for a LGTM here:
https://github.com/kubernetes/kubernetes/pull/78545
thanks.

rajansandeep commented:
@neolit123 @johnbelamaric If the bloated dependency seems like the only problem, I can revert the changes to include [only changes from this commit](https://github.com/kubernetes/kubernetes/pull/78033/commits/e151d73d1f617beb765f1a3cd4984ffc00545555), although it seems like it's all green at @cblecker's end from the dep-team to include all these changes.

Reverting CoreDNS back to 1.3.1 seems like a shortcut to solve this issue.

chrisohaver commented:
FWIW - i'm currently looking into refactoring caddy completely out of the imports (for _this_ PR not for the deps pre-exising this one).

neolit123 commented:
/hold
until this merges at this point: https://github.com/kubernetes/kubernetes/pull/78033

neolit123 commented:
/hold cancel
sorry missed your comment @rajansandeep 

so....it's after midnight for me and i won't babysit either this PR or the revert.
if you think this can merge before the revert, please put the revert on hold using `/hold`.
https://github.com/kubernetes/kubernetes/pull/78545

johnbelamaric commented:
If @cblecker is ok with the deps I am too, we don't need to derail this for that.

chrisohaver commented:
@rajansandeep @cblecker @neolit123 @johnbelamaric 

I refactored the underlying module to have zero deps, @rajansandeep. Can you update the deps here? thanks.

rajansandeep commented:
@neolit123 @johnbelamaric @chrisohaver @fabriziopandini @cblecker 
I have rebased and updated the vendor dep required after @chrisohaver's changes to the migration library.
PTAL
/hold cancel

cblecker commented:
/retest

johnbelamaric commented:
/lgtm

rajansandeep commented:
/retest

cblecker commented:
/test pull-kubernetes-kubemark-e2e-gce-big

neolit123 commented:
/lgtm
but this needs a top level approval now.
@cblecker can you help on this front?

cblecker reviewed (approved):
The new dependencies look fine to me. I haven't reviewed the kubeadm changes, but @neolit123 has and he's an approver for `cmd/kubeadm/`

/assign @liggitt 
for root approval

liggitt commented:
is this still needed with the revert of coredns to 1.3.1? (https://github.com/kubernetes/kubernetes/pull/78545)

rajansandeep commented:
@liggitt The revert of CoreDNS to 1.3.1 was proposed in case this PR doesn't resolve in time.
At this moment, this PR is good to go and we are no longer going to revert CoreDNS to 1.3.1, but continue to have CoreDNS 1.5.0 for k8s 1.15.

https://github.com/kubernetes/kubernetes/pull/78545 will close once this is approved and merged.

liggitt commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -297,6 +298,88 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
+			return errors.Wrap(err, "unable to migrate CoreDNS ConfigMap")
+		}
+
+		if _, err := client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Update(&v1.ConfigMap{
```
what does this do to currently-running CoreDNS pods? do they try to reload the updated/migrated config? if so, will that work?

if the upgrade process stops, does this leave the CoreDNS configmap populated with data that breaks the 1.14-level of CoreDNS?

I expected to leave any 1.14-level configmap alone, so the current CoreDNS pods would continue working, create a 1.15-level configmap and point the 1.15 deployment of CoreDNS at that config

chrisohaver commented:
> what does this do to currently-running CoreDNS pods? do they try to reload the updated/migrated config? if so, will that work?
> if the upgrade process stops, does this leave the CoreDNS configmap populated with data that breaks the 1.14-level of CoreDNS?

CoreDNS will revert back its in-memory prior configuration if it fails to load a new configuration.  The in-memory config will not survive a container restart however.   IOW, if for some reason the upgrade process updates the ConfigMap, but fails to update the Deployment (can this happen?), then CoreDNS would remain running on the old configs, until the CoreDNS containers are restarted, at which point they would fail (until CoreDNS is updated to the correct version).  

> I expected to leave any 1.14-level configmap alone, so the current CoreDNS pods would continue working, create a 1.15-level configmap and point the 1.15 deployment of CoreDNS at that config.

Yes, that would protect against a partial upgrade - if it were to happen.  It would leave some cruft behind each upgrade (the old ConfigMaps).  Would need to decide how to name the configmap (append the CoreDNS version number ok? e.g. "coredns-1.5.0").

@rajansandeep, all thats needed for this would be to create the new ConfigMap under the new name, and update the Deployment to use the new ConfigMap.  Then Kubernetes rolling update takes over from there.

liggitt commented:
> CoreDNS will revert back its in-memory prior configuration if it fails to load a new configuration.

That may be better than crashing, but may not be... presumably they have customized the config for a reason.

> The in-memory config will not survive a container restart however.

So we may be putting the existing deployment in a cannot-restart scenario. That's not good.

> IOW, if for some reason the upgrade process updates the ConfigMap, but fails to update the Deployment (can this happen?), then CoreDNS would remain running on the old configs, until the CoreDNS containers are restarted, at which point they would fail (until CoreDNS is updated to the correct version).

That can definitely happen. ConfigMap update and Deployment update are distinct operations... one can succeed and the following one fail.

> Yes, that would protect against a partial upgrade - if it were to happen. It would leave some cruft behind each upgrade (the old ConfigMaps). Would need to decide how to name the configmap (append the CoreDNS version number ok? e.g. "coredns-1.5.0").
> 
> @rajansandeep, all thats needed for this would be to create the new ConfigMap under the new name, and update the Deployment to use the new ConfigMap. Then Kubernetes rolling update takes over from there.

I'm in favor of doing this in a way that keeps current pods functioning and rolls out the new config safely. I think it's too late in the release to design and test this thoroughly.

I would recommend merging the rollback, designing this config migration and update, ensuring we have test coverage for various scenarios, and getting this in in 1.16

johnbelamaric commented:
There was a recent PR that merged that added the ready plugin to the
config. Was that just for test infra, or for kubeadm installed clusters as
well? Or do we need to roll that back too if we roll back to 1.3?


On Mon, Jun 3, 2019 at 9:31 AM Jordan Liggitt <notifications@github.com>
wrote:

> CoreDNS will revert back its in-memory prior configuration if it fails to
> load a new configuration.
>
> That may be better than crashing, but may not be... presumably they have
> customized the config for a reason.
>
> The in-memory config will not survive a container restart however.
>
> So we may be putting the existing deployment in a cannot-restart scenario.
> That's not good.
>
> IOW, if for some reason the upgrade process updates the ConfigMap, but
> fails to update the Deployment (can this happen?), then CoreDNS would
> remain running on the old configs, until the CoreDNS containers are
> restarted, at which point they would fail (until CoreDNS is updated to the
> correct version).
>
> That can definitely happen. ConfigMap update and Deployment update are
> distinct operations... one can succeed and the following one fail.
>
> Yes, that would protect against a partial upgrade - if it were to happen.
> It would leave some cruft behind each upgrade (the old ConfigMaps). Would
> need to decide how to name the configmap (append the CoreDNS version number
> ok? e.g. "coredns-1.5.0").
>
> @rajansandeep <https://github.com/rajansandeep>, all thats needed for
> this would be to create the new ConfigMap under the new name, and update
> the Deployment to use the new ConfigMap. Then Kubernetes rolling update
> takes over from there.
>
> I'm in favor of doing this in a way that keeps current pods functioning
> and rolls out the new config safely. I think it's too late in the release
> to design and test this thoroughly.
>
> I would recommend merging the rollback, designing this config migration
> and update, ensuring we have test coverage for various scenarios, and
> getting this in in 1.16
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/78033?email_source=notifications&email_token=ACIHRM5FEELPWXGAJFLMHSDPYVBOBA5CNFSM4HNVYRG2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWZ6KTA#issuecomment-498328908>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACIHRMZ25HJ6YENCCSD53BDPYVBOBANCNFSM4HNVYRGQ>
> .
>

rajansandeep commented:
> There was a recent PR that merged that added the ready plugin to the
config. Was that just for test infra, or for kubeadm installed clusters as
well? Or do we need to roll that back too if we roll back to 1.3?

@johnbelamaric the changes are just for kube-up (test-infra) and isn't related to kubeadm.

chrisohaver commented:
Hopefully this will get elevated attention for the next release, and we don't end up in the same situation.

neolit123 commented:
as per the latest comments.
/milestone v1.16

chrisohaver commented:
IMO, all of the related PRs should be reverted...  There were a few.

There were also some tickets closed because we moved to CoreDNS 1.5... those should be reopened.

chrisohaver commented:
I recall also recently seeing a 5k node scalability issue relating to memory exhaustion in CoreDNS after another change, that "went away" after the bump up to CoreDNS 1.5.0.  Don't know if it was coincidence, or if it was causal ... searching ... #78562.

Should keep an eye on that test to make sure it doesn't start failing after CoreDNS revert.

neolit123 commented:
@rajansandeep @chrisohaver do you think we can get these comments to be addressed in 1.16?
https://github.com/kubernetes/kubernetes/pull/78033#issuecomment-498328908

in particular:
> all thats needed for this would be to create the new ConfigMap under the new name, and update the Deployment to use the new ConfigMap. Then Kubernetes rolling update takes over from there.

feature freeze is again around the corner (end of this month).

rajansandeep commented:
@neolit123 yes, I’ll be updating the PR soon. The aim is to get everything included for 1.16.

neolit123 commented:
thanks just ping me to help with reviewers.

rajansandeep commented:
/cc @neolit123 @liggitt 
I have updated the PR to address https://github.com/kubernetes/kubernetes/pull/78033#issuecomment-512582753, PTAL, Thanks!

neolit123 commented:
@rajansandeep 
PTAL at the failing verify test:
```
2019/07/23 21:22:41 ERROR: /home/prow/go/src/k8s.io/kubernetes/build/dependencies.yaml indicates that coredns should be at version 1.3.1, but the following files didn't match:

cmd/kubeadm/app/constants/constants.go

if you are changing the version of coredns, make sure all of the following files are updated with the newest version including /home/prow/go/src/k8s.io/kubernetes/build/dependencies.yaml
then run ./hack/verify-external-dependencies-version.sh
```
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/78033/pull-kubernetes-verify/1153775029968703488

neolit123 reviewed (commented):
thank you for the update @rajansandeep !
i've added quite a few comments but hopefully we can get this resolved soon.
please, let me know if you have any questions based on my comments.

rosti commented on cmd/kubeadm/app/phases/upgrade/postupgrade.go:
```diff
@@ -94,6 +97,17 @@ func PerformPostUpgradeTasks(client clientset.Interface, cfg *kubeadmapi.InitCon
+		errs = append(errs, err)
+	}
+	isMigrationRequired := dns.IsCoreDNSConfigMapMigrationRequired(corefile)
+	if isMigrationRequired {
```
In fact, this whole blob needs to go into `EnsureDNSAddon` or a new func called `UpgradeDNSAddon`.
Even with that in mind, I don't think, that we should backup the CoreDNS configuration into a config map. We may have it locally in a string var, just in case we want to revert to it.

neolit123 replied:
> We may have it locally in a string var, just in case we want to revert to it.

one problem with this approach is that if kubeadm exists the backup will be lost.
if it's stored on disk it is better.

rosti replied:
Indeed a comprehensive crash tolerant upgrade revert system is a good thing to have, but we don't have the time, plus cluster backup tools already provide this sort of functionality.

rosti reviewed (commented):
Thanks for continuing work on this @rajansandeep !

I literally think, that this should be a way simpler change than it should be.
Yes, it's always good to have a backup, but this looks like a precedent for kubeadm. We should be careful and certainly we should not reimplement functionality usually found in cluster backup tools.

rajansandeep commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -298,6 +309,70 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
+		return errors.Wrap(err, "unable to migrate CoreDNS ConfigMap")
+	}
+
+	if _, err := client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Update(&v1.ConfigMap{
```
> we can do this:
> - copy the coredns ConfigMap into coredns-backup
> - point the Deployment at the backup
> - wait for update
> - create a new ConfigMap called coredns that has the new config
> point the Deployment at coredns
> - wait for update
> - if the update is not good point the Deployment back at coredns-backup and copy coredns-backup into coredns.
but it requires two rolling updates...
WDYT?

This is somewhat how it is implemented currently except for the last step. 
If creating a separate configmap for backup is not liked like @rosti pointed out, then I can create a Data backup of the Corefile in the same ConfigMap. 

What I suggest: 

- Copy the `Corefile` data as `Corefile-<version>` as backup. 
- Point the Deployment key to the `Corefile-<version>` data and wait for update.
- Migrate the `Corefile` data which will be compatible with the latest CoreDNS release and 
At this point, the ConfigMap will look as follows:
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: coredns
  namespace: kube-system
data:
  Corefile:  <Migrated Data>
  Corefile-<version>: <Backup Data>
```
- Now point the Deployment key back to `Corefile` and wait for the update to happen successfully.
- If the update is successful, then we can leave it as-is. 
- If not, we will copy the `Corefile-<version>` to the `Corefile`, notifying the user that migration failed. 
As I understand, this will not trigger the changes immediately but it will ensure that the CoreDNS pods will reload on the changes eventually. 

@chrisohaver @neolit123 @rosti let me know if this makes sense.

@rosti, while this might be a little complicated, the requirement that there be a fail-safe route with at least 1 CoreDNS pod running in the case migration fails.
 [This discussion and requirement](https://github.com/kubernetes/kubernetes/pull/78033#issuecomment-498328908) is what prevented it from this change not happening in k8s 1.15.

chrisohaver replied:
Makes sense to me.  This lines up with what we had hashed out on the whiteboard last week.  Is it useful to leave the `Corefile-<version>`in place after the migration is over?  It's really there for fall back if the deployment of the migrated Corefile fails.  Certainly if that step fails, then it's not useful and should be removed. It is possible that the step would pass (CoreDNS not crashing), but leave CoreDNS in an otherwise degraded state (e.g. not behaving as expected for some zone) - in which case it might be useful to an admin to reference the previous version of the config.

neolit123 replied:
from https://github.com/kubernetes/kubernetes/pull/78033#issuecomment-498328908

> all thats needed for this would be to create the new ConfigMap under the new name, and update the Deployment to use the new ConfigMap. Then Kubernetes rolling update takes over from there.

i was under the impression that we should not create backup `Corefile-back` Data entry, but instead backup a whole ConfigMap and point Deployment at that, but OK.

> Point the Deployment key to the Corefile-<version> data and wait for update.

i guess by patching this part?
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/addons/dns/manifests.go#L301-L302

>  Is it useful to leave the Corefile-<version>in place after the migration is over? 

1) if we keep a lot of `Corefile-<version>` entries it can inflate the size of the CM indefinitely.
2) if we keep only one version `Corefile-backup` it will allow rollback to an older version, but not further than that.

i'd go for 2, as kubeadm does not support incremental downgrades.

chrisohaver replied:
Keeping only one previous config makes sense.
We moved to using multiple keys in one map instead of two maps based on feedback (feedback that perhaps we misunderstood).  I think it's just as easy to use separate maps.  I don't have a preference.

rajansandeep replied:
> i was under the impression that we should not create backup Corefile-back Data entry, but instead backup a whole ConfigMap and point Deployment at that, but OK.

I guess the point of the backup was to ensure that it is done in a way that keeps current pods functioning and rolls out the new config safely.

>i guess by patching this part?
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/addons/dns/manifests.go#L301-L302

Yes.

rosti replied:
`Corefile-backup` is good, `Corefile-<version>` will bloat the CM over time.

rajansandeep commented:
> @rajansandeep
> PTAL at the failing verify test:
> 
> ```
> 2019/07/23 21:22:41 ERROR: /home/prow/go/src/k8s.io/kubernetes/build/dependencies.yaml indicates that coredns should be at version 1.3.1, but the following files didn't match:
> 
> cmd/kubeadm/app/constants/constants.go
> 
> if you are changing the version of coredns, make sure all of the following files are updated with the newest version including /home/prow/go/src/k8s.io/kubernetes/build/dependencies.yaml
> then run ./hack/verify-external-dependencies-version.sh
> ```
> 
> https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/78033/pull-kubernetes-verify/1153775029968703488

@neolit123, I did not update the CoreDNS version in kube-up deliberately because I intend to carry out the same CoreDNS migration steps for kube-up upgrade/downgrade as well in https://github.com/kubernetes/kubernetes/pull/78302.

I did not want to repeat what happened in the last release where all upgrade/downgrade tests started to fail after I updated the CoreDNS version and things had to be reverted in the end.

Is there a way to get past this for this time alone?

neolit123 commented:
> @neolit123, I did not update the CoreDNS version in kube-up deliberately because I intend to carry out the same CoreDNS migration steps for kube-up upgrade/downgrade as well in #78302.

i create a issue in k/k that bumping the coredns dependency is currently a bit problematic due to OWNER files. will also not about the kube-up coupling.

cblecker commented:
To the kubeadm folks: which presubmits over kubeadm? Right now I can't see any.

BenTheElder commented:
/test pull-kubernetes-e2e-kind
^ covers some @cblecker 
optional for the moment, also just conformance minus serial tests, but plan to expand that soon... (in progress)

neolit123 commented:
@cblecker 

> To the kubeadm folks: which presubmits over kubeadm? Right now I can't see any.

as @BenTheElder explained `pull-kubernetes-e2e-kind`, but we don't have upgrade presubmits and this PR mainly touches the upgrade logic of kubeadm.

we can catch issues in the periodics https://k8s-testgrid.appspot.com/sig-cluster-lifecycle-kubeadm later, plus i'm probably going to do some local testing in any case.

cblecker commented:
This PR covers the upgrade logic, but it also sets a new default version of CoreDNS for kubeadm, correct? I'm concerned about any other possible regression (such as performance) that we aren't testing by moving kubeadm *before* kube-up.

chrisohaver commented:
> This PR covers the upgrade logic, but it also sets a new default version of CoreDNS for kubeadm, correct? I'm concerned about any other possible regression (such as performance) that we aren't testing by moving kubeadm _before_ kube-up.

Good point. We could decouple the migration logic and the version bump into separate PRs.  I.e. This PR would just introduce the corefile migration function (that in itself is not version specific).  A separate PR would bump the version of CoreDNS to whatever it will be (probably the recently released 1.6.0 at this point, not 1.5.0).

rosti commented:
> > This PR covers the upgrade logic, but it also sets a new default version of CoreDNS for kubeadm, correct? I'm concerned about any other possible regression (such as performance) that we aren't testing by moving kubeadm _before_ kube-up.
> 
> Good point. We could decouple the migration logic and the version bump into separate PRs. I.e. This PR would just introduce the corefile migration function (that in itself is not version specific). A separate PR would bump the version of CoreDNS to whatever it will be (probably the recently released 1.6.0 at this point, not 1.5.0).

I like the idea of decoupling the two in a couple of separate PRs.

rajansandeep committed:
move the corefile backup into a single configmap with a corefile-backup data key

rajansandeep commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -317,8 +318,27 @@ func IsCoreDNSConfigMapMigrationRequired(corefile string) bool {
+	}); err != nil {
+		return errors.Wrap(err, "unable to update the CoreDNS ConfigMap with backup Corefile")
+	}
+	if err := patchCoreDNSDeployment(client, "Corefile-backup"); err != nil {
```
@neolit123 I moved the `Corefile` backup into the same ConfigMap and then patching the deployment to point to `Corefile-backup` here. 
When the CoreDNS deployment is updated to the latest during the upgrade, that step would point the key back to `Corefile`, which will again trigger a rollout. 
Let me know if this logic would work or I should move the "deployment-patch" step into a separate step.

neolit123 replied:
thanks for the update, SGTM!

neolit123 commented:
/remove-sig cli api-machinery cloud-provider node

neolit123 commented:
/remove-area apiserver cloudprovider

neolit123 commented:
/approve
/assign @rosti 

@rosti PTAL and LGTM if you think this is good.
at this point i prefer if we merge the PR, run it trough e2e and iterate if needed before 1.16.

neolit123 commented:
@rajansandeep 
```
if you are changing the version of coredns-kubeadm, make sure all of the following files are updated with the newest version including /home/prow/go/src/k8s.io/kubernetes/build/dependencies.yaml
then run ./hack/verify-external-dependencies-version.sh
```
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/78033/pull-kubernetes-verify/1161403539239473152

it needs an update at this point.

rajansandeep committed:
update build dep

rajansandeep commented:
This still needs root approval for the changes in vendor/deps...

neolit123 commented:
/assign @liggitt 

PTAL WRT approval of the /vendor changes.
we can amend on the configmap/upgrade logic if needed in follow ups.

thanks!

BenTheElder commented on cmd/kubeadm/app/phases/addons/dns/manifests.go:
```diff
@@ -314,9 +314,9 @@ data:
     .:53 {
         errors
         health
+        ready
```
FYI @aojea the ipv6 config may need tweaking.

BenTheElder commented on vendor/github.com/coredns/corefile-migration/migration/caddy/alltokens.go:
```diff
@@ -0,0 +1,33 @@
+// Copyright 2015 Light Code Labs, LLC
```
is this package a copy from a different upstream? do we have accurate licensing info for this?
is there a reason this isn't imported from the original authors? was it modified?

neolit123 replied:
^ @chrisohaver

chrisohaver replied:
Originally I just imported the caddy package. But at the [suggestion](https://github.com/kubernetes/kubernetes/pull/78033#issuecomment-497464374) of reviewers concerned for the number of imported packages, I copied the required pieces locally.  The licensing was copied verbatim from the original project.  IIRC, I had to modify it slightly (a function moved from one package to another), to reduce the amount of copied code.

> do we have accurate licensing info for this (in Godeps/) ?

Sorry, I'm not familiar with the licensing process ... Please let me know what I need to do here to make everyone happy.

BenTheElder replied:
> Originally I just imported the caddy package. But at the suggestion of reviewers concerned for the number of imported packages, I copied the required pieces locally. The licensing was copied verbatim from the original project. 

I see, that makes sense I suppose.

> IIRC, I had to modify it slightly (a function moved from one package to another), to reduce the amount of copied code.

So you modified one of the source files then? Do we need to denote this somehow?

> Sorry, I'm not familiar with the licensing process ... Please let me know what I need to do here to make everyone happy.

I'm not actually sure either, I haven't seen this case before.

chrisohaver replied:
> So you modified one of the source files then? Do we need to denote this somehow?

I recall I had to do that to stop the "bleeding" ... otherwise i'd be duplicating most caddy.  I don't know if we need to or how to properly denote this (comments in the files? readme in same directory?).  IMHO, copying files like this feels a hack to me, and I'd prefer just importing the packages normally.  OTOH, I don't have a good understanding of the concern for importing the package in the first place, other than it created a lot of dependencies (which I infer is bad).

BenTheElder replied:
cc @nikhita @justaugustus

rajansandeep replied:
Is there any resolution on how we should proceed with this so that we can get this PR merged ASAP?

chrisohaver replied:
I propose we just import the packages normally, and remove the copied files.  Any objections?

justaugustus replied:
@chrisohaver -- yeah, I'd feel better about that route.

justaugustus replied:
/hold
(until that's resolved.)

rajansandeep replied:
I fixed the part. @justaugustus PTAL and cancel the hold if you feel it's good. Thanks!

justaugustus replied:
Thanks for fixing the licensing piece, @rajansandeep!
/hold cancel

liggitt commented:
I had the same question w.r.t. that package license

rajansandeep committed:
update dependencies

neolit123 commented:
/lgtm
/approve
/retest

rajansandeep commented:
Looks like all hurdles have been crossed and requires one final approval tag.
/assign @liggitt

liggitt commented:
/approve
